fms/models/mpnet.py

```diff
@@ -0,0 +1,480 @@
+import logging
+from dataclasses import dataclass
+
+import torch
+from torch import nn
+from torch.nn import BCEWithLogitsLoss, CrossEntropyLoss, MSELoss
+
+#from fms.utils.activation import gelu
+from fms.utils.config import ModelConfig
+from fms import models
+from fms.modules.attention import (
+    AttentionKwargs,
+    MultiHeadAttention,
+    get_attention_type,
+)
+from fms.modules.layernorm import LayerNormParameterized
+from fms.modules.feedforward import FeedForwardBlock
+from fms.utils.activation import str_to_activation
+from fms.utils import serialization
+
+logger = logging.getLogger(__name__)
+
+@dataclass
+class MpnetConfig(ModelConfig):
+    src_vocab_size: int = 30_527
+    emb_dim: int = 768
+    nlayers: int = 12
+    nheads: int = 12
+    intermediate_size: int = 3072
+    activation_fn: str = "gelu"
+    hidden_dropout_prob: float = 0.1
+    attention_probs_dropout_prob: float = 0.1
+    max_expected_seq_len: int = 512
+    initializer_range: float = 0.02
+    multiquery_attn: bool = False
+    layer_norm_eps: float = 1e-12
+    hidden_grow_factor: float = 4.0
+    relative_attention_num_buckets: int = 32
+    tie_heads: bool = False
+    pad_id: int = 1
+    bos_token_id: int = 0
+    eos_token_id: int = 2
+    linear_config: Optional[Mapping[str, Any]] = None
+    fused_weights: bool = False
+
+
+class MpnetBlock(nn.Module):
+    def __init__(self, config: MpnetConfig):
+        super().__init__()
+        self.config = config
+        emb_kq = self.config.emb_dim // self.config.nheads
+        emb_v = self.config.emb_dim // self.config.nheads
+
+        kvheads = self.config.nheads
+        #self.ln = LayerNormParameterized(
+        #    self.config.emb_dim,
+        #    elementwise_scale=True,
+        #    elementwise_shift=False,
+        #    use_mean=False,
+        #    eps=self.config.layer_norm_eps,
+        #    use_high_precision_pow=True,
+        #)
+        #self.ff_ln = LayerNormParameterized(
+        #    self.config.emb_dim,
+        #    elementwise_scale=True,
+        #    elementwise_shift=False,
+        #    use_mean=False,
+        #    eps=self.config.layer_norm_eps,
+        #    use_high_precision_pow=True,
+        #)
+        
+        #self.attn = MultiHeadAttention(
+        #    self.config.emb_dim,
+        #    emb_kq,
+        #    emb_v,
+        #    self.config.nheads,
+        #    kvheads,
+        #    p_dropout=self.config.hidden_dropout_prob,
+        #    use_bias=True,
+        #    #position_encoder=rotary_emb,
+        #    fused=self.config.fused_weights,
+        #    linear_config=self.config.linear_config,
+        #)
+        self.attn = MultiHeadAttention(
+            self.config.emb_dim,
+            self.config.emb_dim // self.config.nheads,
+            self.config.emb_dim // self.config.nheads,
+            self.config.nheads,
+            kvheads=1 if self.config.multiquery_attn else self.config.nheads,
```

Comment:
oh if we have multiquery_attn might as well make it kvheads, disregard my previous comment